Python/Product/PythonTools/PythonTools/Extensions.cs

```diff
@@ -826,6 +827,51 @@ public static class Extensions {
             }
         }
 
+        internal static bool IsShellInitialized(this IServiceProvider provider) {
+            bool isInitialized;
+            return provider.TryGetShellProperty((__VSSPROPID)__VSSPROPID4.VSSPROPID_ShellInitialized, out isInitialized) &&
+                isInitialized;
+        }
+
+        class ShellInitializedNotification : IVsShellPropertyEvents {
+            private readonly IVsShell _shell;
+            private readonly uint _cookie;
+            private readonly TaskCompletionSource<object> _tcs;
+
+            public ShellInitializedNotification(IVsShell shell) {
+                _shell = shell;
+                _tcs = new TaskCompletionSource<object>();
+                ErrorHandler.ThrowOnFailure(_shell.AdviseShellPropertyChanges(this, out _cookie));
+            }
+
+            public Task Task => _tcs.Task;
+
+            int IVsShellPropertyEvents.OnShellPropertyChange(int propid, object var) {
+                if (propid == (int)__VSSPROPID4.VSSPROPID_ShellInitialized) {
+                    if (var is bool && (bool)var) {
+                        _shell.UnadviseShellPropertyChanges(_cookie);
+                        _tcs.SetResult(null);
+                    }
+                } else if (propid == (int)__VSSPROPID6.VSSPROPID_ShutdownStarted) {
+                    _shell.UnadviseShellPropertyChanges(_cookie);
+                    _tcs.TrySetCanceled();
+                }
+                return VSConstants.S_OK;
+            }
+        }
+
+        internal static Task WaitForShellInitializedAsync(this IServiceProvider provider) {
+            if (provider.IsShellInitialized()) {
+                return Task.FromResult<object>(null);
+            }
+            return new ShellInitializedNotification(provider.GetShell()).Task;
```

Comment:
Is there a risk of waiting forever here if the shell initializes after you've checked, but before you register for the property changes event?